core/csv_logger.py

```diff
@@ -0,0 +1,127 @@
+import csv
+import re
+import threading
+import time
+from pathlib import Path
+from typing import TYPE_CHECKING, Union, Mapping
+
+_stdlock = threading.Lock()
+_filelock = threading.Lock()
+
+
+if TYPE_CHECKING:
+    from scm.plams.core.basejob import Job
+
+
+class LoggerCSV:
+    """basic logger that mimic the behavior of plams.log with adding option of personalize the file logging location"""
+
+    def __init__(self, log_path="logfile.csv", stdout_level=3, date=True, time=True, init: bool = True):
+        self.log_path = Path(log_path)
+        self.stdout_level = stdout_level
+        self.date = date
+        self.time = time
+        self.init = init
+
+    def log(self, message: Mapping[str, Union[str, int, float]], level=7):
+        """Log a dictionary message in CSV format with verbosity level.
+
+        Logs are printed independently to the CSV logfile and to the standard output based on configuration levels.
+        Date and/or time can be added based on configuration. All logging activity is/should thread safe.
+
+        :param message: The message to log as a dictionary.
+        :type message: dict
+        :param level: 1: important; 3: normal; 5: verbose; 7: debug, defaults to 0
+        :type level: int, optional
+        """
+        if not self.init:
+            if level <= 3:
+                with _stdlock:
+                    print(str(message))
+            return
+
+        current_headers = get_current_headers(self.log_path)
+        new_headers_needed = False
+
+        # Prepare row and check headers
+        row = []
+        headers = current_headers.copy()
+        for key in message:
+            if key not in headers:
+                headers.append(key)
+                new_headers_needed = True
+            row.append(message.get(key))
+        extra_headers_needed = new_headers_needed and (current_headers != [])
+
+        if extra_headers_needed:
+            raise NotImplementedError(
+                f"found new headers! {current_headers=} BUT {headers=}, this option is not currently supported"
+            )
```

Comment:
Right, because otherwise you would have to re-write the file with the missing `,,,`...

Could be an argument for json lines as you mentioned, as this would be more flexible, but might be less easier for users to interact with than a csv file.


For csv though, should the headers be inferred, or just set.

As in, should this be derived class called `JobCSVLogger` and you just call `log(job)` 🤔